### PR TITLE
Pgxmock batch example in Go

### PR DESCRIPTION
**Pull request by giuliocn**
This PR would like to improve my work on #201

### CODE REVIEW

**Comment by pashagolub** [comment]
Why didn't you just reopen the previous pull request?

File: examples/batch/batch.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	return err
+}
+
+func requestBatch(db PgxIface) (err error) {
+
+	// Initialize a database object
+	tx, err := db.Begin(context.Background())
+	if err != nil {
+		return fmt.Errorf("requestBatch: %s", err)
+	}
```
Comment (pashagolub) [inline examples/batch/batch.go:75]: Why do you think we need to start a new transaction here? Batch is executed within a single transaction anyway. Do you plan to add other statements?

File: examples/batch/batch_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+func TestShouldRollbackOnFailure(t *testing.T) {
+	mock, err := pgxmock.NewPool()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer mock.Close()
+
+	mock.ExpectBegin()
+	// TODO: mock.ExpectBatch()
+	mock.ExpectRollback()
```
Comment (pashagolub) [inline examples/batch/batch_test.go:30]: What do you mean `TODO`? It's already there for several months, I believe

**Review by pashagolub — CHANGES_REQUESTED** [review]

**Comment by giuliocn** [comment]
> Why didn't you just reopen the previous pull request?

@pashagolub I have no right to re-open closed PR in this repository. I cannot see the button near "Comment".

**Comment by giuliocn** [comment]
The example could be completed by serving all the results within an HTML page. What Do you think?

**Comment by pashagolub** [comment]
Sorry for delay with answer.

I want examples to be as real as possible showing the real life use cases. That why I don't buy this one:
```go
	example.batch.Queue("SELECT * FROM ledger")
	example.batch.Queue("SELECT * FROM ledger WHERE amount = 1")
```

I will never in my life execute such a batch. Why? The first result set has it all.

**Comment by pashagolub** [comment]
Tests failed and coverage is below threshold.

```sh

=== RUN   TestExpectBatch
--- PASS: TestExpectBatch (0.00s)
=== RUN   TestExpectBegin
    .\pgxmock\examples\batch\batch_test.go:50: there were unfulfilled expectations: there is a remaining expectation which was not matched: ExpectedBegin => expecting call to Begin() or to BeginTx()
--- FAIL: TestExpectBegin (0.00s)
FAIL
coverage: 6.7% of statements
FAIL	github.com/pashagolub/pgxmock/v4/examples/batch	0.029s
```

File: examples/batch/batch.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	output strings.Builder
+}
+
+func (ex *ExampleBatch) insertRow(fields ...string) (err error) {
+	for i, value := range fields {
+		fields[i] = "$$" + value + "$$"
+	}
+	sql := `INSERT INTO metadata (title, authors, subject, description)
+	VALUES 
+	(` + strings.Join(fields[:4], ", ") + ");"
```
Comment (pashagolub) [inline examples/batch/batch.go:35]: We don't do parameters concatenation, this is the possible place for SQL injection. We should use parameterized queries.

File: examples/batch/batch.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+		db:    db,
+		batch: &pgx.Batch{},
+	}
+	if err = example.databaseSetup(); err != nil {
+		panic(err)
+	}
+	defer example.Close()
+
+	// Add an example to database table
+	example.insertRow(
```
Comment (pashagolub) [inline examples/batch/batch.go:141]: Why we insert here and in `databaseSetup()`?

File: examples/batch/batch.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	if ex.output.Len() == 0 {
+		ex.output = temp
+	}
+	fmt.Fprint(w, ex.output.String())
+
+}
+
+func (ex *ExampleBatch) requestBatch() (err error) {
```
Comment (pashagolub) [inline examples/batch/batch.go:98]: I don't see any need in this function. Let's keep it simple

File: examples/batch/batch.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+type ExampleBatch struct {
+	batch  *pgx.Batch
+	br     pgx.BatchResults
+	db     PgxPoolInterface
+	output strings.Builder
+}
+
+func (ex *ExampleBatch) insertRow(fields ...string) (err error) {
```
Comment (pashagolub) [inline examples/batch/batch.go:29]: I don't see any value in `insertRow` function. And it's not covered by tests. It's better to remove it and keep things simple

File: examples/batch/batch.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+	"strings"
+
+	pgx "github.com/jackc/pgx/v5"
+	pgconn "github.com/jackc/pgx/v5/pgconn"
+	pgxpool "github.com/jackc/pgx/v5/pgxpool"
+)
+
+type PgxPoolInterface interface {
+	Begin(context.Context) (pgx.Tx, error)
+	Exec(context.Context, string, ...interface{}) (pgconn.CommandTag, error)
```
Comment (pashagolub) [inline examples/batch/batch.go:17]: Nor `Begin` neither `Exec` used anywhere in code. Although this is not an error, we want to show our user that interfaces must be minimal

**Review by pashagolub — CHANGES_REQUESTED** [review]

**Comment by giuliocn** [comment]
@pashagolub Can you explain in what order a single row is retrieved with `BatchResults`? Can I consume a `Rows` object independently for each query? 

Thanks for your comments. 💡

**Comment by pashagolub** [comment]
Yes, this is the idea behind the batch. You send all queries at once and then receive all results at once. And you are free to work with every result set as you wish.

If the example below the magic of sending and execution happened during `SendBatch()`. And then you just work with results you got back firing `results.Exec()`. **This is not you really execute** something, it's just to tell pgx you treat all statements as `Exec()`. But, of course, you can change it to `Query()` or even `QueryRow()`, see [here](https://github.com/jackc/pgx/blob/v5.6.0/batch.go#L80).

```go
func (pg *postgres) BulkInsertUsers(ctx context.Context, users []model.User) error {
  query := `INSERT INTO users (name, email) VALUES (@userName, @userEmail)`

  batch := &pgx.Batch{}
  for _, user := range users {
    args := pgx.NamedArgs{
      "userName": user.Name,
      "userEmail": user.Email,
    }
    batch.Queue(query, args)
  }
  
  results := pg.db.SendBatch(ctx, batch)
  defer results.Close()

  for _, user := range users {
    _, err := results.Exec()
    if err != nil {
      var pgErr *pgconn.PgError
      if errors.As(err, &pgErr) && pgErr.Code == pgerrcode.UniqueViolation {
          log.Printf("user %s already exists", user.Name)
          continue
      }

      return fmt.Errorf("unable to insert row: %w", err)
    }
  }

  return results.Close()
}
```

File: examples/batch/batch.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+		{`The Art of Baking: A Beginner's Guide`, `Emily Baker`, `Cooking, Baking, Food`, `Learn the fundamentals of baking and create delicious treats from scratch. Master techniques like kneading dough, frosting cakes, and piping cookies.`},
+		{`The History of Ancient Rome`, `David Johnson`, `History, Ancient Rome, Civilization`, `Delve into the fascinating history of the Roman Empire. Explore its rise to power, its cultural achievements, and its eventual decline.`},
+	})
+
+	// Send a custom batch operation to database
+	example.SendCustomBatch([]string{
+		"SELECT title FROM metadata",
+		"SELECT authors FROM metadata",
+		"SELECT subject, description FROM metadata",
+	})
```
Comment (pashagolub) [inline examples/batch/batch.go:144]: I don't buy it. It should be one query
```sql
SELECT title, authors, subject, description FROM metadata
```

This is bad example.

File: examples/batch/batch.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	// Print batch results
+	for _, query := range example.batch.QueuedQueries {
+		fmt.Println(query.SQL)
+		rows, _ := example.results.Query()
+		for rows.Next() {
+			values, _ := rows.Values()
+			fmt.Println(values)
+		}
+	}
+}
```
Comment (pashagolub) [inline examples/batch/batch.go:155]: Useless piece of code. No purpose. No meaning.

File: examples/batch/batch.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+	Exec(context.Context, string, ...interface{}) (pgconn.CommandTag, error)
+	SendBatch(context.Context, *pgx.Batch) pgx.BatchResults
+	Close()
+}
+
+type exampleBatch struct {
+	batch   *pgx.Batch       // batch pointer
+	results pgx.BatchResults // query results
+	db      PgxPoolInterface // database connection
+}
```
Comment (pashagolub) [inline examples/batch/batch.go:23]: I don't understand the meaning of this.

**Comment by pashagolub** [comment]
Unfortunately, I can't accept this work. This is not a proper example of using `pgx.Batch`. 
Thanks for your time.

**Pull request closed without merging**